### PR TITLE
SDA-7166 Update the model for node_pools endpoints

### PR DESCRIPTION
**Pull request by lucasponce**
It's my first PR on this repo.

The existing "node_pools" endpoints are not exposed to the public model, and this is a prerequisite step before we can connect them with the ROSA CLI.

cc @pvasant @tbrisker 

Please, feel free to suggest if this flow is correct or if I should send this PR with other changes in the SDK and/or another repo.

Thanks,
Lucas

### CODE REVIEW

File: model/clusters_mgmt/v1/aws_node_pool_type.model [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
+Copyright (c) 2021 Red Hat, Inc.
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/aws_node_pool_type.model:2]: what year?

File: model/clusters_mgmt/v1/aws_node_pool_type.model [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// Representation of aws node pool specific parameters.
+class AWSNodePool {
+    // InstanceType is an ec2 instance type for node instances (e.g. m5.large).
+    InstanceType String
+
+    // InstanceProfile is the AWS EC2 instance profile, which is a container for an IAM role that the EC2 instance uses.
+    InstanceProfile String
+
+	// Optional keys and values that the installer will add as tags to all AWS resources it creates
+	Tags [String]String
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/aws_node_pool_type.model:26]: indentation looks off here

File: model/clusters_mgmt/v1/node_pool_autoscaling_type.model [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+limitations under the License.
+*/
+
+// Representation of a autoscaling in a node pool.
+class NodePoolAutoscaling {
+    // The minimum number of replicas for the node pool.
+	Min Integer
+
+    // The maximum number of replicas for the node pool.
+	Max Integer
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/node_pool_autoscaling_type.model:23]: Indentation here too

File: model/clusters_mgmt/v1/node_pool_type.model [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	// The number of Machines (and Nodes) to create.
+	// Replicas and autoscaling cannot be used together.    
+	Replicas Integer
+
+	// Details for auto-scaling the machine pool.
+	// Replicas and autoscaling cannot be used together.    
+	Autoscaling NodePoolAutoscaling
+
+    // Specifies whether health checks should be enabled for machines in the NodePool.
+    AutoRepair Boolean
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/node_pool_type.model:31]: indentation, 

This is the bane of my existance with this repo, the indentation between ide's and this `model` syntax

File: model/clusters_mgmt/v1/node_pool_type.model [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Representation of a node pool in a cluster.
+class NodePool {
+	// ID used to identify the cluster that this machinepool is attached to.
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/node_pool_type.model:19]: Is this correct `ID` :thinking:

File: model/clusters_mgmt/v1/node_pools_resource.model [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Manages the collection of node pools of a cluster.
+resource NodePools {
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/node_pools_resource.model:18]: I think this might need a locator, normally a root resource for all resources :thinking:

**Comment by lucasponce** [comment]
Thanks, @ciaranRoche; I think I've fixed most of the comments. 
Would you like to give it another pass?
Thx.

File: model/clusters_mgmt/v1/cluster_type.model [hunk 7]
```diff
@@ -158,6 +158,9 @@ class Cluster {
 	// List of machine pools on this cluster.
 	link MachinePools []MachinePool
 
+    // List of node pools on this cluster
+    link NodePools []NodePool
```
Comment (ciaranRoche) [inline model/clusters_mgmt/v1/cluster_type.model:162]: out of curiosity what is a node pool? and how is it different to machine pools?

Comment (lucasponce) [reply model/clusters_mgmt/v1/cluster_type.model:162]: It's a similar concept but for Hosted clusters.
These NodePools are "detached" from the control plane.
From OCM, the original work created different endpoints for them.

But from ROSA CLI we may invoke one or another depending on hosted cluster or not, that would be my next task and I'm enabling this in the model as a prior step (it wasn't updated before).

Comment (ciaranRoche) [reply model/clusters_mgmt/v1/cluster_type.model:162]: okay makes sense, can we add that to the comments, as the comments are used to generate the docs in the openapi. Be good to call out the difference between node pools and machine pools.

**Comment by lucasponce** [comment]
Updated, tested locally with an example like:

```
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/openshift-online/ocm-sdk-go"
	v1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
)

func main() {
	logger, err := sdk.NewGoLoggerBuilder().
		Debug(true).
		Build()
	if err != nil {
		log.Fatalf("can't build logger: %v\n", err)
	}

	connection, err := sdk.NewConnectionBuilder().
		Logger(logger).
		Tokens(os.Getenv("OFFLINE_ACCESS_TOKEN")).
		URL("http://localhost:9000").
		Insecure(true).
		Build()
	if err != nil {
		log.Fatalf("Can't build connection: %v", err)
	}
	defer connection.Close()

	clusterId := "20756lq9aha19chgodjd8krdtr49q8o6"
	clusterGetResp, err := connection.ClustersMgmt().V1().Clusters().Cluster(clusterId).Get().Send()
	if err != nil {
		log.Fatalf("could not find cluster : %v", err)
	}

	fmt.Println("got cluster - ", clusterGetResp.Body().Name())

	nodePoolsListResp, err := connection.ClustersMgmt().V1().Clusters().Cluster(clusterId).NodePools().List().Send()
	if err != nil {
		log.Fatalf("could not find node pools for a cluster : %v", err)
	}

	fmt.Println("got list - ", nodePoolsListResp.Items().Len())

	nodePool, err := v1.NewNodePool().Replicas(3).Build()
	if err != nil {
		log.Fatalf("could not build nodepool : %v", err)
	}
	nodePoolAddResp, err := connection.ClustersMgmt().V1().Clusters().Cluster(clusterId).NodePools().Add().Body(nodePool).Send()
	if err != nil {
		log.Fatalf("could not add a node pool for a cluster : %v", err)
	}

	fmt.Println("got new nodepool - ", nodePoolAddResp.Body().HREF())
}
```

Once that the SDK is enabled and updated from main CS code, it can enable more end-to-end tests on this area.

**Comment by lucasponce** [comment]
I have a local build of the "ocm-sdk-go" that works locally with the rosa CLI changes.
Once this PR is reviewed+merged, should I update the "ocm-sdk-go"?
How are the updates of the new "ocm-sdk-go" version propagated into ROSA or any other client? Is that done automatically?

File: model/clusters_mgmt/v1/node_pool_autoscaling_type.model [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
Comment (pvasant) [inline model/clusters_mgmt/v1/node_pool_autoscaling_type.model:2]: year:)

**Comment by tbrisker** [comment]
@lucasponce once this is merged, a new release with it must be made, then the sdk needs to be updated to use the new version, regenerated, and released, and then any consumer can update the version they use. (yes, i know, this is way too long of a process for every api change)

File: model/clusters_mgmt/v1/node_pool_autoscaling_type.model (hunk 8 above)
Comment (lucasponce) [reply model/clusters_mgmt/v1/node_pool_autoscaling_type.model:2]: +1

File: model/clusters_mgmt/v1/node_pool_resource.model [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
Comment (lucasponce) [inline model/clusters_mgmt/v1/node_pool_resource.model:2]: Bad year

File: model/clusters_mgmt/v1/node_pool_type.model [hunk 10]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
Comment (lucasponce) [inline model/clusters_mgmt/v1/node_pool_type.model:2]: bad year

File: model/clusters_mgmt/v1/node_pools_resource.model [hunk 11]
```diff
@@ -0,0 +1,45 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
Comment (lucasponce) [inline model/clusters_mgmt/v1/node_pools_resource.model:2]: year

**Comment by lucasponce** [comment]
Thanks, @tbrisker.
I've updated the PR with the minor fixes reviewed by @ciaranRoche and @pvasant.
I think it's ready for merging.

**Review by tbrisker — APPROVED** [review]
Thanks @lucasponce , i'll tag it for you

**Comment by tbrisker** [comment]
https://github.com/openshift-online/ocm-api-model/actions/runs/3558592802

**Pull request merged**